cdb2api/cdb2api.c

```diff
@@ -5872,16 +5874,115 @@ static int set_up_ssl_params(cdb2_hndl_tp *hndl)
+}
+#endif /* OPENSSL_NO_DEPRECATED */
+
+/* For OpenSSL < 1.0.0. */
+static unsigned long ssl_threadid_deprecated()
+{
+    return (unsigned long)pthread_self();
+}
+
+static void cdb2_ssl_lock(int mode, int type, const char *file, int line)
+{
+    int rc;
+    if (mode & CRYPTO_LOCK) {
+        if ((rc = pthread_mutex_lock(&cdb2_ssl_locks[type])) != 0)
+            my_ssl_eprintln("Failed to lock pthread mutex: rc = %d.", rc);
+    } else {
+        if ((rc = pthread_mutex_unlock(&cdb2_ssl_locks[type])) != 0)
+            my_ssl_eprintln("Failed to unlock pthread mutex: rc = %d.", rc);
+    }
+}
+#endif /* CRYPTO_num_locks */
+
+static int ssl_init(void)
+{
+#ifndef OPENSSL_THREADS
+    /* OpenSSL is not configured for threaded applications. */
+    ssl_sfeprint(NULL, 0, my_ssl_eprintln, "OpenSSL is not configured with thread support.");
+    return EPERM;
+#endif /* OPENSSL_THREADS */
+
+#ifndef CRYPTO_num_locks
+    /* Configure SSL locking.
+       This is only required for OpenSSL < 1.1.0. */
+    int nlocks = CRYPTO_num_locks();
+    cdb2_ssl_locks = malloc(sizeof(pthread_mutex_t) * nlocks);
+    if (cdb2_ssl_locks == NULL) {
+        ssl_sfeprint(NULL, 0, my_ssl_eprintln, "Failed to allocate SSL locks.");
+        return ENOMEM;
+    }
+
+    int rc = 0;
+    for (int ii = 0; ii < nlocks; ++ii) {
+        if ((rc = pthread_mutex_init(&cdb2_ssl_locks[ii], NULL)) != 0) {
+            /* Whoops - roll back all we have done. */
+            while (ii > 0) {
+                --ii;
+                pthread_mutex_destroy(&cdb2_ssl_locks[ii]);
+            }
+            free(cdb2_ssl_locks);
+            cdb2_ssl_locks = NULL;
+            ssl_sfeprint(NULL, 0, my_ssl_eprintln, "Failed to initialize mutex: %s", strerror(rc));
+            return rc;
+        }
+    }
+#ifdef OPENSSL_NO_DEPRECATED
+    CRYPTO_THREADID_set_callback(ssl_threadid);
+#else
+    /* Use deprecated functions for OpenSSL < 1.0.0. */
+    CRYPTO_set_id_callback(ssl_threadid_deprecated);
+#endif /* OPENSSL_NO_DEPRECATED */
+    CRYPTO_set_locking_callback(cdb2_ssl_lock);
+#endif /* CRYPTO_num_locks */
+
+    /* Configure the library. */
+#if OPENSSL_VERSION_NUMBER >= 0x10100000L
+    OPENSSL_init_ssl(OPENSSL_INIT_LOAD_CONFIG, NULL);
+#else
+    OPENSSL_config(NULL);
+    SSL_library_init();
+    SSL_load_error_strings();
+#endif /* OPENSSL_VERSION_NUMBER */
+    return 0;
+}
+
+/*
+ * Initialize SSL library.
+ *
+ * PARAMETERS
+ * init_libssl     - set to non-zero to initialize libssl
+ * init_libcrypto  - set to non-zero to initialize libcrypto
+ * err             - set to NULL to print to stderr
+ * n               - length of the error string buffer
+ *
+ * RETURN VALUES
+ * 0 upon success
+ */
 static int cdb2_called_ssl_init = 0;
-pthread_mutex_t fend_ssl_init_lock = PTHREAD_MUTEX_INITIALIZER;
-int cdb2_init_ssl(int init_libssl, int init_libcrypto)
+static pthread_mutex_t fend_ssl_init_lock = PTHREAD_MUTEX_INITIALIZER;
+int cdb2_init_ssl(void)
```

Comment:
You would want to keep these 2 arguments. An application may initialize libssl or libcrypto outside cdb2api. And these 2 arguments give the application the ability to prevent double initialization of the libraries. See the doc [here](https://bloomberg.github.io/comdb2/c_api.html#cdb2_init_ssl).

Reply:
Yes!